File: magellan-library/src/test/java/com/wealthfront/magellan/transitions/CrossfadeTransitionTest.java
package com.wealthfront.magellan.transitions;

import android.view.View;

import com.wealthfront.magellan.Direction;
import com.wealthfront.magellan.transitions.MagellanTransition.Callback;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static androidx.test.core.app.ApplicationProvider.getApplicationContext;
import static com.google.common.truth.Truth.assertThat;
import static com.wealthfront.magellan.Direction.BACKWARD;
import static com.wealthfront.magellan.Direction.FORWARD;
import static org.robolectric.Robolectric.flushForegroundThreadScheduler;
import static org.robolectric.Robolectric.getForegroundThreadScheduler;

@RunWith(RobolectricTestRunner.class)
public class CrossfadeTransitionTest {

  boolean onAnimationEndCalled;
  View from = new View(getApplicationContext());
  View to = new View(getApplicationContext());

  @Before
  public void setUp() {
    onAnimationEndCalled = false;
    from.setVisibility(VISIBLE);
    to.setVisibility(GONE);
    getForegroundThreadScheduler().pause();
  }

  @Test
  public void animateShow() {
    checkAnimate(FORWARD);
  }

  @Test
  public void animateHide() {
    checkAnimate(BACKWARD);
  }

  private void checkAnimate(Direction direction) {
    new CrossfadeTransition().animate(from, to, direction, new Callback() {
      @Override
      public void onAnimationEnd() {
        onAnimationEndCalled = true;
      }
    });
    flushForegroundThreadScheduler();
    assertThat(onAnimationEndCalled).isTrue();
    assertThat(from.getVisibility()).isEqualTo(GONE);
    assertThat(to.getVisibility()).isEqualTo(VISIBLE);
  }

}

File: magellan-library/src/test/java/com/wealthfront/magellan/transitions/CrossfadeTransitionTest.kt
package com.wealthfront.magellan.transitions

import android.os.Looper.getMainLooper
import android.view.View
import androidx.test.core.app.ApplicationProvider.getApplicationContext
import com.google.common.truth.Truth.assertThat
import com.wealthfront.magellan.Direction
import com.wealthfront.magellan.Direction.BACKWARD
import com.wealthfront.magellan.Direction.FORWARD
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows.shadowOf

@RunWith(RobolectricTestRunner::class)
class CrossfadeTransitionTest {

  var onAnimationEndCalled = false
  var from = View(getApplicationContext())
  var to = View(getApplicationContext())

  @Before
  fun setUp() {
    onAnimationEndCalled = false
    from.visibility = View.VISIBLE
    to.visibility = View.GONE
  }

  @Test
  fun animateShow() {
    animate(FORWARD)
    shadowOf(getMainLooper()).idle()
    assertThat(onAnimationEndCalled).isTrue()
    assertThat(from.visibility).isEqualTo(View.GONE)
    assertThat(to.visibility).isEqualTo(View.VISIBLE)
  }

  @Test
  fun animateHide() {
    animate(BACKWARD)
    shadowOf(getMainLooper()).idle()
    assertThat(onAnimationEndCalled).isTrue()
    assertThat(from.visibility).isEqualTo(View.GONE)
    assertThat(to.visibility).isEqualTo(View.VISIBLE)
  }

  @Test
  fun interrupt() {
    val transition = animate(FORWARD)
    transition.interrupt()
    assertThat(onAnimationEndCalled).isTrue()
    assertThat(from.visibility).isEqualTo(View.GONE)
    assertThat(to.visibility).isEqualTo(View.VISIBLE)
  }

  private fun animate(direction: Direction): MagellanTransition {
    return CrossfadeTransition().apply {
      animate(from, to, direction) {
        onAnimationEndCalled = true
      }
    }
  }
}
